linkerd/io/src/peek.rs

```diff
@@ -0,0 +1,48 @@
+use crate::{AsyncRead, AsyncWrite, PrefixedIo};
+use bytes::BytesMut;
+use futures::{try_ready, Future, Poll};
+use std::io;
+
+/// A future of when some `Peek` fulfills with some bytes.
+#[derive(Debug)]
+pub struct Peek<T>(Option<Inner<T>>);
+
+#[derive(Debug)]
+struct Inner<T> {
+    buf: BytesMut,
+    io: T,
+}
+
+// === impl Peek ===
+
+impl<T: AsyncRead + AsyncWrite> Peek<T> {
+    pub fn with_capacity(capacity: usize, io: T) -> Self
+    where
+        Self: Sized + Future,
+    {
+        let buf = BytesMut::with_capacity(capacity);
+        Peek(Some(Inner { buf, io }))
+    }
+}
+
+impl<T: AsyncRead + AsyncWrite> Future for Peek<T> {
+    type Item = PrefixedIo<T>;
```

Comment:
Note that `Peek` has changed to be a concrete type that wraps an `Io` to produce a `PrefixedIo`